 for whole team).nk</h3>

In [1]:
from io import StringIO

In [4]:
import requests, zipfile
r = requests.get('https://archive.ics.uci.edu/static/public/222/bank+marketing.zip')

In [36]:
import numpy as np

In [2]:
import pandas as pd
df = pd.read_csv('https://archive.ics.uci.edu/static/public/222/data.csv')

In [5]:
df.head()

age           job  marital  education default  balance housing loan  \
0   58    management  married   tertiary      no     2143     yes   no   
1   44    technician   single  secondary      no       29     yes   no   
2   33  entrepreneur  married  secondary      no        2     yes  yes   
3   47   blue-collar  married        NaN      no     1506     yes   no   
4   33           NaN   single        NaN      no        1      no   no   

  contact  day_of_week month  duration  campaign  pdays  previous poutcome   y  
0     NaN            5   may       261         1     -1         0      NaN  no  
1     NaN            5   may       151         1     -1         0      NaN  no  
2     NaN            5   may        76         1     -1         0      NaN  no  
3     NaN            5   may        92         1     -1         0      NaN  no  
4     NaN            5   may       198         1     -1         0      NaN  no

In [4]:
df.shape

(45211, 17)

In [20]:
print(f"Raw data has {df.shape[0]} rows, and   {df.shape[1]} columns")

Raw data has 45211 rows, and   17 columns


In [22]:
df['poutcome'].unique()

array([nan, 'failure', 'other', 'success'], dtype=object)

In [7]:
from ucimlrepo import fetch_ucirepo 
  


In [8]:
# fetch dataset 
bank_marketing = fetch_ucirepo(id=222) 
  
# data (as pandas dataframes) 
X = bank_marketing.data.features 
y = bank_marketing.data.targets 
  
# metadata 
print(bank_marketing.metadata) 

{'uci_id': 222, 'name': 'Bank Marketing', 'repository_url': 'https://archive.ics.uci.edu/dataset/222/bank+marketing', 'data_url': 'https://archive.ics.uci.edu/static/public/222/data.csv', 'abstract': 'The data is related with direct marketing campaigns (phone calls) of a Portuguese banking institution. The classification goal is to predict if the client will subscribe a term deposit (variable y).', 'area': 'Business', 'tasks': ['Classification'], 'characteristics': ['Multivariate'], 'num_instances': 45211, 'num_features': 16, 'feature_types': ['Categorical', 'Integer'], 'demographics': ['Age', 'Occupation', 'Marital Status', 'Education Level'], 'target_col': ['y'], 'index_col': None, 'has_missing_values': 'yes', 'missing_values_symbol': 'NaN', 'year_of_dataset_creation': 2014, 'last_updated': 'Fri Aug 18 2023', 'dataset_doi': '10.24432/C5K306', 'creators': ['S. Moro', 'P. Rita', 'P. Cortez'], 'intro_paper': {'title': 'A data-driven approach to predict the success of bank telemarketing'

In [9]:
# variable information 
print(bank_marketing.variables) 

           name     role         type      demographic  \
0           age  Feature      Integer              Age   
1           job  Feature  Categorical       Occupation   
2       marital  Feature  Categorical   Marital Status   
3     education  Feature  Categorical  Education Level   
4       default  Feature       Binary             None   
5       balance  Feature      Integer             None   
6       housing  Feature       Binary             None   
7          loan  Feature       Binary             None   
8       contact  Feature  Categorical             None   
9   day_of_week  Feature         Date             None   
10        month  Feature         Date             None   
11     duration  Feature      Integer             None   
12     campaign  Feature      Integer             None   
13        pdays  Feature      Integer             None   
14     previous  Feature      Integer             None   
15     poutcome  Feature  Categorical             None   
16            

In [10]:
df['contact'].unique()

array([nan, 'cellular', 'telephone'], dtype=object)

In [11]:
df['pdays'].unique()

array([ -1, 151, 166,  91,  86, 143, 147,  89, 140, 176, 101, 174, 170,
       167, 195, 165, 129, 188, 196, 172, 118, 119, 104, 171, 117, 164,
       132, 131, 123, 159, 186, 111, 115, 116, 173, 178, 110, 152,  96,
       103, 150, 175, 193, 181, 185, 154, 145, 138, 126, 180, 109, 158,
       168,  97, 182, 127, 130, 194, 125, 105, 102,  26, 179,  28, 183,
       155, 112, 120, 137, 124, 187, 190, 113, 162, 134, 169, 189,   8,
       144, 191, 184, 177,   5,  99, 133,  93,  92,  10, 100, 156, 198,
       106, 153, 146, 128,   7, 121, 160, 107,  90,  27, 197, 136, 139,
       122, 157, 149, 135,  30, 114,  98, 192, 163,  34,  95, 141,  31,
       199,  94, 108,  29, 268, 247, 253, 226, 244, 239, 245, 204, 231,
       238, 258, 230, 254, 265,  71, 223, 246, 250, 266, 240, 205, 261,
       259, 241, 260, 234, 251, 225, 161, 237, 262, 248, 255, 220, 227,
       206, 224, 249, 235, 228, 263,   2, 270, 232, 252, 207, 200, 269,
       233, 256, 273, 272, 242, 264, 208, 214, 222, 271, 203, 22

In [13]:
#to check if all the columns in a row is null.
df[['age','job','marital','education','default','balance','housing','loan','contact','day_of_week','month','duration','campaign','pdays','previous','poutcome']].isnull().apply(lambda x: all(x), axis=1) 

0        False
1        False
2        False
3        False
4        False
         ...  
45206    False
45207    False
45208    False
45209    False
45210    False
Length: 45211, dtype: bool

In [15]:
df.columns[df.isna().any()].tolist() #to list the columns with atleast 1 null value

['job', 'education', 'contact', 'poutcome']

In [16]:
df['job'].unique()

array(['management', 'technician', 'entrepreneur', 'blue-collar', nan,
       'retired', 'admin.', 'services', 'self-employed', 'unemployed',
       'housemaid', 'student'], dtype=object)

In [17]:
df['education'].unique()

array(['tertiary', 'secondary', nan, 'primary'], dtype=object)

In [18]:
df['contact'].unique()

array([nan, 'cellular', 'telephone'], dtype=object)

In [19]:
df['poutcome'].unique()

array([nan, 'failure', 'other', 'success'], dtype=object)

In [21]:
#Number of Missing values in each column
print(df.isna().sum())

age                0
job              288
marital            0
education       1857
default            0
balance            0
housing            0
loan               0
contact        13020
day_of_week        0
month              0
duration           0
campaign           0
pdays              0
previous           0
poutcome       36959
y                  0
dtype: int64


<h3>Correlation Matrix:</h3>

In [22]:
corr = df.corr()
plt.subplots(figsize=(15,10))
sns.heatmap(corr, xticklabels=corr.columns,yticklabels=corr.columns, annot=True, )

ValueError: could not convert string to float: 'management'

We can see here, there is a problem with the data. Let's check the datatypes of the features first.

In [23]:
df.dtypes

age             int64
job            object
marital        object
education      object
default        object
balance         int64
housing        object
loan           object
contact        object
day_of_week     int64
month          object
duration        int64
campaign        int64
pdays           int64
previous        int64
poutcome       object
y              object
dtype: object

<H3>Outliers</H3>
<br>
We will generate some summary statistics for the numeric features now as this is the quickest way to help us determine if our dataset has any outliers.

We will first separate the numeric and categorical features.

In [37]:
train_num = df.select_dtypes(include=[np.number]) #select numeric columns into a new dataframe

train_cat = df.select_dtypes(include=['object'])  #store categorical features into another dataframe

In [40]:
train_num.columns

Index(['age', 'balance', 'day_of_week', 'duration', 'campaign', 'pdays',
       'previous'],
      dtype='object')

In [41]:
train_cat.columns

Index(['job', 'marital', 'education', 'default', 'housing', 'loan', 'contact',
       'month', 'poutcome', 'y'],
      dtype='object')

In [42]:
train_num.head()

age  balance  day_of_week  duration  campaign  pdays  previous
0   58     2143            5       261         1     -1         0
1   44       29            5       151         1     -1         0
2   33        2            5        76         1     -1         0
3   47     1506            5        92         1     -1         0
4   33        1            5       198         1     -1         0

<h3>Check for missing values</h3>

We are aware, the missing values are present in only 4 categorical features: job, education, contact, and poutcome, with number of missing values 288, 1857,  13020, 36959.

We won't delete the entire columns with the missing values as at this point we are not sure of the importance of each feature in predicting the target yet.

We can try an approach here by deleting an entire row with null values, for job as it has only 288 null values as compared to the actual size of the data 45211 records.

<h4>Approach 1: Delete row with missing values for "job" feature </h4>me       36959

In [50]:
df_del_rows_traincat = train_cat.copy()
df_del_rows_traincat.head()

job  marital  education default housing loan contact month  \
0    management  married   tertiary      no     yes   no     NaN   may   
1    technician   single  secondary      no     yes   no     NaN   may   
2  entrepreneur  married  secondary      no     yes  yes     NaN   may   
3   blue-collar  married        NaN      no     yes   no     NaN   may   
4           NaN   single        NaN      no      no   no     NaN   may   

  poutcome   y  
0      NaN  no  
1      NaN  no  
2      NaN  no  
3      NaN  no  
4      NaN  no

In [60]:
df_del_rows_traincat.dropna(subset = ['job'], inplace = True)
df_del_rows_traincat.head()


job  marital  education default housing loan contact month  \
0    management  married   tertiary      no     yes   no     NaN   may   
1    technician   single  secondary      no     yes   no     NaN   may   
2  entrepreneur  married  secondary      no     yes  yes     NaN   may   
3   blue-collar  married        NaN      no     yes   no     NaN   may   
5    management  married   tertiary      no     yes   no     NaN   may   

  poutcome   y  
0      NaN  no  
1      NaN  no  
2      NaN  no  
3      NaN  no  
5      NaN  no

In [61]:
df_del_rows_traincat.columns[df_del_rows_traincat.isna().any()].tolist() #to list the columns with atleast 1 null value

['education', 'contact', 'poutcome']

In [62]:
df_del_rows_traincat

job   marital  education default housing loan    contact  \
0        management   married   tertiary      no     yes   no        NaN   
1        technician    single  secondary      no     yes   no        NaN   
2      entrepreneur   married  secondary      no     yes  yes        NaN   
3       blue-collar   married        NaN      no     yes   no        NaN   
5        management   married   tertiary      no     yes   no        NaN   
...             ...       ...        ...     ...     ...  ...        ...   
45206    technician   married   tertiary      no      no   no   cellular   
45207       retired  divorced    primary      no      no   no   cellular   
45208       retired   married  secondary      no      no   no   cellular   
45209   blue-collar   married  secondary      no      no   no  telephone   
45210  entrepreneur   married  secondary      no      no   no   cellular   

      month poutcome    y  
0       may      NaN   no  
1       may      NaN   no  
2       may      NaN   no  
3       may      NaN   no  
5       may      NaN   no  
...     ...      ...  ...  
45206   nov      NaN  yes  
45207   nov      NaN  yes  
45208   nov  success  yes  
45209   nov      NaN   no  
45210   nov    other   no  

[44923 rows x 10 columns]

<h3>Approach 2 : Imputation by most frequent values (Mode)</h3>
<br>
In this method, you can impute the missing or NULL values by the most frequent or common value in the column. The advantage of this method is that it can work with both numerical and categorical features. You can use Scikit-learn’s SimpleImputer class for this imputation method. This class also works for imputation by mean and median approaches as well.

In [59]:
# impute Graduated and Family_Size features with most_frequent values
from sklearn.impute import SimpleImputer
impute_mode = SimpleImputer(strategy = 'most_frequent')
impute_mode.fit(df[['education', 'contact', 'poutcome']])

df_del_rows_traincat[['education', 'contact', 'poutcome']] = impute_mode.transform(df[['education', 'contact', 'poutcome']])


ValueError: Length of values (45211) does not match length of index (44923)

<h3> Outliers for Numeric Values </h3>

<h4>Using Z score </h4>

In [43]:
def checkZscore(train_num,column_name):
    
    mean   = train_num[column_name].mean()
    stn_dev = train_num[column_name].std()
    
    upper_band = mean + 3*stn_dev
    lower_band = mean - 3*stn_dev
    
    print('\n\n' + "mean of : " + str(column_name) + " is " + str(mean) +
          '\n' + "Standard deviation of : " + str(column_name) + " is " + str(stn_dev) +
          '\n' + "Upper and Lower Limits of " + str(column_name) + " is "  + str((lower_band,upper_band))
         )
    
    return upper_band , lower_band

In [44]:
print('Outlier infos of the features:')
for column_name in train_num.columns :
    upper_band , lower_band = checkZscore(outlier_check_cols , column_name)
    outliers = train_num[(train_num[column_name] < lower_band) | (train_num[column_name] > upper_band)][column_name].values
    print('\n' + str(column_name) + " has " + str(len(outliers)) + " outliers : "    + str(outliers))
    print('\n\n**********************************************************\n')

Outlier infos of the features:


mean of : age is 40.93621021432837
Standard deviation of : age is 10.618762040975431
Upper and Lower Limits of age is (9.079924091402077, 72.79249633725466)

age has 381 outliers : [83 75 75 83 75 76 83 85 90 82 73 74 75 78 85 80 78 73 75 94 83 79 78 83
 73 76 83 77 73 74 73 77 80 82 75 77 78 78 86 76 74 77 74 79 74 78 73 73
 95 74 78 75 75 82 74 81 76 75 77 73 85 83 78 73 79 80 78 79 82 81 79 79
 79 73 75 75 78 75 80 80 74 73 77 79 73 80 83 73 78 81 81 74 75 82 80 76
 74 74 77 77 76 74 74 73 82 77 79 89 76 81 74 73 77 84 86 76 73 74 82 83
 73 73 74 78 86 76 79 81 77 82 73 73 74 73 76 83 80 79 77 74 84 95 77 78
 79 83 83 73 77 74 83 76 73 81 80 75 74 75 77 83 81 73 74 80 76 77 74 87
 76 92 78 82 78 76 73 77 76 73 82 81 80 76 74 75 80 84 87 76 78 73 78 81
 76 79 73 82 77 75 76 81 77 75 74 75 78 77 83 73 81 80 80 86 79 79 87 81
 75 84 78 89 73 85 87 80 81 85 79 76 74 92 77 74 81 73 76 80 82 77 76 78
 77 74 76 77 80 76 76 73 84 79 77 75 76 75 75 75 77 82 8

<h5><i> We can see that the outliers are present in age : 381 , balance : 745, duration : 963, campaign : 840, pdays : 1723 , previous : 582 </i></h5>